keps/sig-node/4381-dra-numeric-parameters/README.md

```diff
@@ -0,0 +1,1080 @@
+spec:
+  count: 10
+  env:
+    HELLO_WORLD: "Thanks for the fish."
+```
+
+### Notes/Constraints/Caveats
+
+For this proposal to work, kube-scheduler must have permission to read, list and
+watch arbitrary resources. This is necessary because it is not known in advance
+which CRDs will be used by DRA drivers.
+
+This may change at runtime as classes get added or removed. To avoid repeated
+GET operations for those CRDs, dynamic informers will have to be used that get
+started and stopped as needed.
+
+### Risks and Mitigations
+
+Granting kube-scheduler permission to read arbitrary types could be used by an
+attacker to extract information. However, for that an attacker first would have
+to gain control of the scheduler, get it to execute API calls, and then
+retrieve the result. This seems like a low risk.
+
+Using a single NodeResourceCapacity object per node is simple and convenient
+also for admins, but might run into size limitations. This should not be a
+problem in practice because the number of drivers per node is expected to be
+low and the capacity information from each should be small. Limits for the
+JSON fragment can be introduced to prevent excessive sizes.
+
+## Design Details
+
+### Decoding claim parameters
+
+Extracting the numeric parameters can be done without knowing the CRD:
+
+- Look up the resource based on group and kind with a REST mapper.
+
+- Retrieve the object with a `dynamic.Client` as an `unstructured.Unstructured`
+  instance (basically a nested map).
+
+- Retrieve the sub-map via the JSON path.
+
+- Re-encode that part as JSON with the normal `encode/json` package.  In the
+  example above, this will include `kind` and `apiVersion:
+  counter.dra.config.k8s.io` because those are the defaults and only allowed
+  values.
+
+- Decode into the numeric parameter with the apimachinery JSON decoder,
+  ignoring unknown fields like `env` in this example.
+
+### ResourceClass extension
+
+A new, optional field in ResourceClass enables numeric parameters for claims
+using this class:
+
+```
+type ResourceClass struct {
+    ...
+
+    // If (and only if) allocation of claims using this class is handled
+    // via numeric parameters, then NumericParameters must list all claim
+    // parameter types that are allowed for claims.
+    NumericParameters []NumericParameterType
+}
+
+type NumericParameterType struct {
+	// APIGroup is the group for the resource being referenced. It is
+	// empty for the core API. This matches the group in the APIVersion
+	// that is used when creating the resources.
+	// +optional
+	APIGroup string
+
+	// Kind is the type of resource being referenced. This is the same
+	// value as in the parameter object's metadata.
+	Kind string
+
+	// FieldPath is a dot separarated JSON path that defines where in the
+	// parameter object the numeric parameters are embedded.
+	FieldPath string
+
+	// Shareable is copied into the AllocationResult when a claim
+	// gets allocated.
+	Shareable bool
```

Comment:
If I remember correctly, this is currently at the top-level (i.e. as part of a `ResourceClass` object). It's probably worth mentioning that it is moved inside of here.

Reply:
The ResourceClass has no `Shareable`. Such a field is only in [`AllocationResult`](https://github.com/kubernetes/kubernetes/blob/922fbbf5b70d461118dc03e4eb2eb2ddc8762481/staging/src/k8s.io/api/resource/v1alpha2/types.go#L169). The scheduler needs to know what to put there, which is why the information needs be provided in `NumericParameterType`.

Reply:
I am not sure why this would be in the parameter type. Are getting at that a given parameter allocation can be exclusive or non-exclusive? I think that probably belongs in the actual semantic model class (or maybe it is satisfiable with a single-counter attribute, I am not sure and need to think more on it...the "mount for monitoring" case comes to mind).

Reply:
"mount for monitoring" is different than simply being able to mark a given claim as sharable or not.

In the common case, when a resource is bound to a claim it cannot also be bound to another claim at the same time. Meaning, that the resource is removed from the pool of available resources that can be assigned to a claim once it has been allocated. The `sharable` field in this case indicates whether the claim that resource is bound to can be referenced by multiple pods or not. 

The case of "mount for monitoring" is different, because you want to create a claim that can inject a resource into a container *without* removing it from the pool of available resources.

Since all allocation decisions are made by individual DRA drivers in the original DRA design, supporting claims that provide these "monitoring" or "management" capabilities, is left up to the DRA driver itself. In the DRA driver for GPUs, for example, we accomplish this as follows:
1. Provide a special `ClaimParameters` object called `GpuManagementClaimParameters` (which have the same selector capabilities as a normal `GpuClaimParameters` object). 
2. Pods that reference claims that (in turn) reference a `GpuManagementClaimParameters` object will be granted access to the GPUs selected by it, but those GPUs will not be removed from the pool of GPUs available for other claims.
3. We enforce that only pods running in a namespace with a `management.gpu.nvidia.com=true` annotation applied are allowed to reference claims with `GpuManagementClaimParameters`. Pods not running in such a namespace will not be granted access to such a claim and will fail to start.

Given that the new plan of record is to outsource the translation of vendor specific `ClaimParamaters` to a DRA controller, it seems feasible that the controller for our DRA driver for GPUs could continue to do this namespace check (and other vendors could do something else, if desired).

However, *something* must exist in the type that a controller translates its parameters into, that allows it to specify that the resource selected should *not* be accounted against the total number of available resources.

Reply:
> I am not sure why this would be in the parameter type. Are getting at that a given parameter allocation can be exclusive or non-exclusive?

A claim can be exclusive or non-exclusive. Exclusive (= sharable false) means that only a single pod may reserve it at a time. This is enforced.

Putting this into the in-tree claim parameters also work. It doesn't need to be in the class.

Reply:
Moved into ResourceClaimParameters where it's controlled by the vendor or user (if we allow users to create them).